zebra/zebra_tc.c

```diff
@@ -0,0 +1,475 @@
+	} else {
+		new = hash_get(zrouter.qdisc_hash, qdisc,
+			       tc_qdisc_alloc_intern);
+		(void)dplane_tc_qdisc_install(new);
+	}
+}
+
+void zebra_tc_qdisc_uninstall(struct zebra_tc_qdisc *qdisc)
+{
+	if (IS_ZEBRA_DEBUG_TC)
+		zlog_debug("%s: uninstall tc qdisc ifindex %d kind %s",
+			   __func__, qdisc->qdisc.ifindex,
+			   tc_qdisc_kind2str(qdisc->qdisc.kind));
+
+	(void)dplane_tc_qdisc_uninstall(qdisc);
+
+	if (tc_qdisc_release(qdisc, true))
+		zlog_debug("%s: tc qdisc being deleted we know nothing about",
+			   __func__);
+}
+
+uint32_t zebra_tc_class_hash_key(const void *arg)
+{
+	const struct zebra_tc_class *class;
+	uint32_t key;
+
+	class = arg;
+
+	key = jhash_2words(class->class.ifindex, class->class.handle, 0);
+
+	return key;
+}
+
+bool zebra_tc_class_hash_equal(const void *arg1, const void *arg2)
+{
+	const struct zebra_tc_class *c1, *c2;
+
+	c1 = (const struct zebra_tc_class *)arg1;
+	c2 = (const struct zebra_tc_class *)arg2;
+
+	if (c1->class.ifindex != c2->class.ifindex)
+		return false;
+
+	if (c1->class.handle != c2->class.handle)
+		return false;
+
+	return true;
+}
+
+static void *tc_class_alloc_intern(void *arg)
+{
+	struct zebra_tc_class *class;
+	struct zebra_tc_class *new;
+
+	class = (struct zebra_tc_class *)arg;
+
+	new = XCALLOC(MTYPE_TMP, sizeof(*new));
+
+	memcpy(new, class, sizeof(*class));
+
+	return new;
+}
+
+static struct zebra_tc_class *tc_class_free(struct zebra_tc_class *hash_data,
+					    bool free_data)
+{
+	hash_release(zrouter.class_hash, hash_data);
+
+	if (free_data) {
+		XFREE(MTYPE_TMP, hash_data);
+		return NULL;
+	}
+
+	return hash_data;
+}
+
+static struct zebra_tc_class *tc_class_release(struct zebra_tc_class *class,
+					       bool free_data)
+{
+	struct zebra_tc_class *lookup;
+
+	lookup = hash_lookup(zrouter.class_hash, class);
+
+	if (!lookup)
+		return NULL;
+
+	return tc_class_free(lookup, free_data);
+}
+
+void zebra_tc_class_add(struct zebra_tc_class *class)
+{
+	if (IS_ZEBRA_DEBUG_TC)
+		zlog_debug(
+			"%s: add tc class ifindex %d handle %04x:%04x kind %s",
+			__func__, class->class.ifindex,
+			(class->class.handle & 0xffff0000u) >> 16,
+			class->class.handle & 0x0000ffffu,
+			tc_qdisc_kind2str(class->class.kind));
+
+	struct zebra_tc_class *found;
+	struct zebra_tc_class *new;
+
+	/*
+	 * We find the class in the hash by (ifindex, handle) directly, and by
+	 * testing their deep equality to seek out whether it's an update.
+	 */
+	found = hash_lookup(zrouter.class_hash, class);
+
+	if (found) {
+		/*
+		 * In fact the deep equal is not even necessary, since updating
+		 * the same contents via dplane would do nothing harmful in
+		 * terms of TCLASS.
+		 */
+
+		/*
+		 * old = tc_class_release(found, false);
+		 * (void)dplane_tc_class_delete(old);
+		 * new = hash_get(zrouter.class_hash, class, ...);
+		 * (void)dplane_tc_class_add(new);
+		 */
+
+		/* currently a direct update is enough */
+
+		new = hash_get(zrouter.class_hash, class,
+			       tc_class_alloc_intern);
```

Comment:
This in-place update suggests that not all fields relevant to class uniqueness relative to the `class_hash` are being hashed as part of the key function - otherwise, why would we need to re-insert the class that we have just found in the hash table?

Can you explain what this code is doing?

Reply:
The `hash_key` for `class` only involves its `ifindex` and `handle` (also holds true for `hash_equal`), since this is what the kernel treated as its uniqueness. 

Suppose that the hash already contains `class(ifindex=0, handle=0xa, rate=40Mbit)` and we want to add `class'(ifindex=0, handle=0xa, rate=60Mbit)`.
Finding `class'` in hash (a class where `ifindex = class'.ifindex` and `handle = class'.handle`) will get us `class(ifindex=0, handle=0xa, rate=40Mbit)`, then we know it's an update; if we cannot find then it's a brand-new one. To sync this update, we inserted it so `class` is replaced by `class'` in the hash.